datalad/support/gitrepo.py

```diff
@@ -3123,85 +3170,137 @@ def _diffstatus_get_state_props(self, f, from_state, to_state,
-            # for subdatasets leave the 'modified' judgement to the caller
-            # for supporting corner cases, such as adjusted branch
-            # which require inspection of a subdataset
-            # TODO we could have a new file that is already staged
-            # but had subsequent modifications done to it that are
-            # unstaged. Such file would presently show up as 'added'
-            # ATM I think this is OK, but worth stating...
-            state = ('modified'
-                     if against_commit or to_state['type'] != 'dataset'
-                     else None
-                    ) if f.exists() or f.is_symlink() else 'deleted'
-            # TODO record before and after state for diff-like use
-            # cases
+        state = self._diffstatus_get_state(
+            f, from_state, to_state,
+            modified_in_worktree,
+            against_commit, eval_submodule_state,
+        )
+
+        # compile properties of the diff state
+        # TODO sort out
+        props = {}
+
+        to_sha = to_state.get('gitshasum') if to_state else None
+        from_sha = from_state.get('gitshasum') if from_state else None
 
         if state in ('clean', 'added', 'modified', None):
             # assign present gitsha to any record
             # state==None can only happen for subdatasets that
             # already existed, so also assign a sha for them
-            props['gitshasum'] = to_sha
+            if to_sha:
+                # with a typechange there would be no gitsha
+                # for the new content, despite a known modification
+                props['gitshasum'] = to_sha
             if 'bytesize' in to_state:
                 # if we got this cheap, report it
                 props['bytesize'] = to_state['bytesize']
             elif state == 'clean' and 'bytesize' in from_state:
                 # no change, we can take this old size info
                 props['bytesize'] = from_state['bytesize']
+
         if state in ('clean', 'modified', 'deleted', None):
             # assign previous gitsha to any record
             # state==None can only happen for subdatasets that
             # already existed, so also assign a sha for them
             props['prev_gitshasum'] = from_sha
+
+        # current type reporting
+        if modified_in_worktree is not False:
+            # we have a report of a modified type, include if
+            # not None (i.e. vanished)
+            if modified_in_worktree is not None:
+                props['type'] = modified_in_worktree
+        elif to_state and 'type' in to_state:
+            props['type'] = to_state['type']
+
+        if state == 'modified':
+            # for modifications we want to report types for both states
+            props['prev_type'] = from_state['type']
+
+        if state == 'deleted':
+            # report the type that was deleted
+            props['type'] = from_state['type']
```

Comment:
shouldn't "current" state just be None since it was deleted?

Reply:
Absolutely! https://github.com/datalad/datalad/pull/6878/commits/3fa23fcd288af6714f451c18c03d8e19aa152694 in https://github.com/datalad/datalad/pull/6878